src/Random.mo

```diff
@@ -0,0 +1,303 @@
+/// as the IC, is intricate. Some basic rules need to be followed by the
+/// user of this module to obtain (and maintain) the benefits of crypto-
+/// graphic randomness:
+/// - cryptographic entropy (randomness source) is only obtainable
+///   asyncronously in discrete chunks of 256 bits (32-byte sized `Blob`s)
+/// - all bets must be closed *before* entropy is being asked for in
+///   order to decide them
+/// - this implies that the same entropy (i.e. `Blob`) - or surplus entropy
+///   not utilised yet - cannot be used for a new round of bets without
+///   losing the cryptographic guarantees.
+///
+/// Concretely, the below classes `Finite` and `Cyclic`, as well as the
+/// `*From` methods risk the carrying-over of state from previous rounds.
+/// These are provided for performance (and convenience) reasons, and need
+/// special care when used. Similar caveats apply for user-defined (pseudo)
+/// random number generators.
+
+import Prim "mo:prim";
+import P "Prelude"
+
+module {
+
+  /// Drawing from a finite supply of entropy `Finite` provides
+  /// methods to obtain random values. When the entropy is used up,
+  /// `null` is returned. Otherwise the outcomes' distributions are
+  /// stated for each method.
+  public class Finite(entropy : Blob) {
+    let it : { next : () -> ?Word8 } = entropy.bytes();
+
+    /// Uniformly distributes outcomes in the numeric range [0 .. 255].
+    public func byte() : ?Nat8 {
+      switch (it.next()) {
+        case (?w) ?Prim.word8ToNat8 w;
+        case null null
+      }
+    };
+
+    /// Bool iterator splitting up a byte of entropy into 8 bits
+    let bit : { next : () -> ?Bool } = {
+      var mask = 0x80 : Word8;
+      var byte = 0x00 : Word8;
+      next = func () : ?Bool {
+        if (0 : Word8 == mask) {
+          switch (it.next()) {
+            case null null;
+            case (?w) {
+              byte := w;
+              mask := 0x40;
+              ?(0 : Word8 != byte & (0x80 : Word8))
+            }
+          }
+        } else {
+          let m = mask;
+          mask >>= (1 : Word8);
+          ?(0 : Word8 != byte & m)
+        }
+      }
+    };
+
+    /// Simulates a coin toss. Both outcomes have equal probability.
+    public func coin() : ?Bool {
+      bit.next()
+    };
+
+    /// Uniformly distributes outcomes in the numeric range [0 .. 2^n - 1].
+    public func range(p : Nat8) : ?Nat {
+      var pp = p;
+      var acc : Nat = 0;
+      for (i in it) {
+        if (8 : Nat8 <= pp)
+        { acc := acc * 256 + Prim.word8ToNat(i) }
+        else if (0 : Nat8 == pp)
+        { return ?acc }
+        else {
+          acc *= Prim.word8ToNat(1 << Prim.nat8ToWord8 pp);
+          let mask : Word8 = -1 >> Prim.nat8ToWord8(8 - pp);
+          return ?(acc + Prim.word8ToNat(i & mask))
+        };
+        pp -= 8
+      };
+      null
+    };
+
+    /// Counts the number of heads in `n` fair coin tosses.
+    public func binomial(n : Nat8) : ?Nat8 {
```

Comment:
Can we share some of this definition with `binomialFrom`?

Reply:
a3eaf38.

Reply:
Unfortunately I'd have to perform unpleasant massaging of the result to get out an optional. That would complicate the refactor. I'll leave this as is. Wake me up when we have an unobtrusive monad syntax.